federatedscope/core/trainers/context.py

```diff
@@ -182,3 +217,116 @@ def reset_used_dataset(self):
         self.cur_data_splits_used_by_routine.pop()
         self.cur_data_split = self.cur_data_splits_used_by_routine[-1] if \
             len(self.cur_data_splits_used_by_routine) != 0 else None
+
+    def clear(self, lifecycle):
+        """Clear the variables at the end of their lifecycle
+
+        Args:
+            lifecycle: the type of lifecycle
+
+        Returns:
+
+        """
+        attrs = list(self.lifecycles[lifecycle])
+        # Clear general attributes
+        for var in attrs:
+            if hasattr(self[var], "clear"):
+                self[var].clear()
+            else:
+                del self[var]
+            self.lifecycles[lifecycle].remove(var)
+        # Also clear the attributes under the current mode
+        self.var.clear(lifecycle)
+
+
+class BasicCtxVar(object):
+    """Basic variable class
+
+    Arguments:
+        lifecycle: specific lifecycle of the attribute
+    """
+
+    LIEFTCYCLES = ["batch", "epoch", "routine", None]
+
+    def __init__(self, lifecycle=None):
+        assert lifecycle in BasicCtxVar.LIEFTCYCLES
+
+        self.lifecycle = lifecycle
+
+
+class CtxReferVar(BasicCtxVar):
+    """To store the reference variables with specific lifecycle and clear function, e.g. model, data, dataloader
+
+    Arguments:
+        obj: the stored obj
+        lifecycle: the specific lifecycle of the variable
+        efunc: a function that will be called when the lifecycle ends
+
+    """
+    def __init__(self, obj, lifecycle=None, efunc=None):
+        super(CtxReferVar, self).__init__(lifecycle=lifecycle)
+        self.obj = obj
+        self.efunc = efunc
+
+    def clear(self):
+        if self.efunc is not None:
+            self.efunc(self.obj)
+
+
+def CtxStatsVar(init=0., lifecycle="routine"):
```

Comment:
It seems that the `CtxStatsrVar` is a particular case of `CtxReferVar`, however, these two classes may confuse users to deciding which `XxxVar` for certain variables we need to use. Besides, the dynamic type specialization of `CtxStatsrVar` is a bit tricky. 
We can simply provide a `CtxVar` that contains any object with easy-to-use space-time management in terms of cur_mode/cur_dataset and lifecycles.